JuliaStream.jl/src/CUDAStream.jl

```diff
@@ -0,0 +1,154 @@
+include("Stream.jl")
+using CUDA
+
+const CuData = StreamData{T,CuArray{T}} where {T}
+const TBSize = 1024::Int
+const DotBlocks = 256::Int
+
+function devices()::Vector{DeviceWithRepr}
+  return !CUDA.functional(false) ? String[] :
+         map(d -> (d, "$(CUDA.name(d)) ($(repr(d)))", "CUDA.jl"), CUDA.devices())
+end
+
+function make_stream(
+  arraysize::Int,
+  scalar::T,
+  device::DeviceWithRepr,
+  silent::Bool,
+)::Tuple{CuData{T},Nothing} where {T}
+
+  if arraysize % TBSize != 0
+    error("arraysize ($(arraysize)) must be divisible by $(TBSize)!")
+  end
+
+  # so CUDA's device is 0 indexed, so -1 from Julia
+  CUDA.device!(device[1])
+  selected = CUDA.device()
+  # show_reason is set to true here so it dumps CUDA info 
+  # for us regardless of whether it's functional
+  if !CUDA.functional(true)
+    error("Non-functional CUDA configuration")
+  end
+  if !silent
+    println("Using CUDA device: $(CUDA.name(selected)) ($(repr(selected)))")
+    println("Kernel parameters: <<<$(arraysize ÷ TBSize),$(TBSize)>>>")
+  end
+  return (
+    CuData{T}(
+      CuArray{T}(undef, arraysize),
+      CuArray{T}(undef, arraysize),
+      CuArray{T}(undef, arraysize),
+      scalar,
+      arraysize,
+    ),
+    nothing,
+  )
+end
+
+function init_arrays!(data::CuData{T}, _, init::Tuple{T,T,T}) where {T}
+  CUDA.fill!(data.a, init[1])
+  CUDA.fill!(data.b, init[2])
+  CUDA.fill!(data.c, init[3])
+end
+
+function copy!(data::CuData{T}, _) where {T}
+  function kernel(a, c)
+    i = (blockIdx().x - 1) * blockDim().x + threadIdx().x # only blockIdx starts at 1
+    @inbounds c[i] = a[i]
+    return
+  end
+  @cuda blocks = data.size ÷ TBSize threads = TBSize kernel(data.a, data.c)
+  CUDA.synchronize()
+end
+
+function mul!(data::CuData{T}, _) where {T}
+  function kernel(b, c, scalar)
+    i = (blockIdx().x - 1) * blockDim().x + threadIdx().x # only blockIdx starts at 1
+    @inbounds b[i] = scalar * c[i]
+    return
+  end
+  @cuda blocks = data.size ÷ TBSize threads = TBSize kernel(data.b, data.c, data.scalar)
+  CUDA.synchronize()
```

Comment:
I'm not familiar with the underlying architecture here, but do you really need to synchronize after every kernel? That's not great for performance. Our GPU-to-CPU copy already synchronizes, so typically you don't need anything else.

Reply:
The original implementation issues `cudaDeviceSynchronize();` after kernel submission (see https://github.com/UoB-HPC/BabelStream/blob/8f9ca7baa77c874897e1691c895729a39d959012/CUDAStream.cu#L161) so this port must match that. Does the `synchronize` in Julia does a lot more than `cudaDeviceSynchronize()`?  From what I can see, synchronisation doesn't seem to add any more overhead compare to the native version.

Reply:
It should have similar overhead.